### ISSUE TITLE
Performance drop due to syncthreads

### ISSUE THREAD

**Issue by gchauras** [halide/Halide#614]
Can you tell me where you insert syncthreads in the generated code for GPU targets. I want to test the performance drop because of these. There are situations where the pure def is undef() but syncthreads are still injected, and elsewhere there are two consecutive syncthreads. These are unnecessary but I don't know if they have any impact at all.

**Comment by abadams** [comment]
Look for the class InjectThreadBarriers in FuseGPUThreadLoops.cpp. E.g. you
could comment out line 444.

**Comment by zvookin** [comment]
Barriers are injected in FuseGPUThreadLoops.cpp .

-Z-

**Comment by gchauras** [comment]
Disabling all syncthreads gives a 8% performance bump. I have several update defs including some left undef() to resolve bank conflicts. Currently, barriers are injected even after pure or update defs which are left undef(). I am wondering if I can remove these, even though this won't help much. How can I detect that a Pipeline or Block contains undef(). I can see that RemoveUndef.cpp: has 

```
if (!produce.defined()) {
  "Produce step elided due to undef()"
}
```
My wishful thinking led to try the following unsuccessfully of course in FuseGPUThreadLoops.cpp:46
```
produce = Block::make(produce, (produce.defined() ? barrier() : 0));
```

**Comment by abadams** [comment]
We represent no-ops as constant strings or integers (for anonymous ones).
Try something like this:

bool is_noop(Stmt s) {
  const Evaluate *e = s.as<Evaluate>();
  return e && is_const(e->value);
}

**Comment by gchauras** [comment]
I couldn't get this done but I am abandoning this idea. Thanks anyway.

By the way, I noticed that if you leave a pure def undef(), do not specify any scheduling for the pure def, and specify gpu_threads() etc for update defs, then the buffer is not allocated on device. It seems its copied from the host. This makes sense because the pure def's schedule does not tell the compiler that the buffer must be allocated on the device. So when it sees the update def, it scrambles for a buffer and brings it from the host.

I generally pre-upload all Buffers<> to the GPU - those that contain filter weights etc. I noticed that all these were uploaded on each run if the there was no scheduling for the undef-ed pure defs. This went away once I specified gpu_single_thread() as pure defs scheduling.

I am hazy on details -- but I can send a test case later if you think this is a bug.

**Comment by abadams** [comment]
Hrm, I bet fuse_gpu_thread_loops is viewing the undef as a cpu-side usage.
I'd consider this a bug. The fix may be as simple as reordering the two
lowering stages remove_undef_values and fuse_gpu_thread_loops. They are
already side-by-side.

**Comment by gchauras** [comment]
The order you mentioned didn't change the result  but I tried the following order which fixed the issue. I guess the problem was that host to dev copying has happened before fuse_gpu_trheads or remove_undef could kick in.
```
remove_undef()
inject_host_dev_buffer_copies()
fuse_gpu_thread_loops()
```

**Comment by abadams** [comment]
Should be fixed by #621
